In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("../data/raw/retail_store_inventory.csv")

df.shape

(73100, 15)

In [3]:
df.head()

,Date,Store ID,Product ID,Category,Region,Inventory Level,Units Sold,Units Ordered,Demand Forecast,Price,Discount,Weather Condition,Holiday/Promotion,Competitor Pricing,Seasonality
0,2022-01-01,S001,P0001,Groceries,North,231,127,55,135.47,33.50,20,Rainy,0,29.69,Autumn
1,2022-01-01,S001,P0002,Toys,South,204,150,66,144.04,63.01,20,Sunny,0,66.16,Autumn
2,2022-01-01,S001,P0003,Toys,West,102,65,51,74.02,27.99,10,Sunny,1,31.32,Summer
3,2022-01-01,S001,P0004,Toys,North,469,61,164,62.18,32.72,10,Cloudy,1,34.74,Autumn
4,2022-01-01,S001,P0005,Electronics,East,166,14,135,9.26,73.64,0,Sunny,0,68.95,Summer


In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 73100 entries, 0 to 73099
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Date                73100 non-null  object 
 1   Store ID            73100 non-null  object 
 2   Product ID          73100 non-null  object 
 3   Category            73100 non-null  object 
 4   Region              73100 non-null  object 
 5   Inventory Level     73100 non-null  int64  
 6   Units Sold          73100 non-null  int64  
 7   Units Ordered       73100 non-null  int64  
 8   Demand Forecast     73100 non-null  float64
 9   Price               73100 non-null  float64
 10  Discount            73100 non-null  int64  
 11  Weather Condition   73100 non-null  object 
 12  Holiday/Promotion   73100 non-null  int64  
 13  Competitor Pricing  73100 non-null  float64
 14  Seasonality         73100 non-null  object 
dtypes: float64(3), int64(5), object(7)
memory usage: 8.4+

In [5]:
df.describe(include="all").T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
Date,73100,731,2022-01-01,100,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Store ID,73100,5,S001,14620,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Product ID,73100,20,P0001,3655,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Category,73100,5,Furniture,14699,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Region,73100,4,East,18349,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Inventory Level,73100.0,NaN,NaN,NaN,274.469877,129.949514,50.0,162.0,273.0,387.0,500.0
Units Sold,73100.0,NaN,NaN,NaN,136.46487,108.919406,0.0,49.0,107.0,203.0,499.0
Units Ordered,73100.0,NaN,NaN,NaN,110.004473,52.277448,20.0,65.0,110.0,155.0,200.0
Demand Forecast,73100.0,NaN,NaN,NaN,141.49472,109.254076,-9.99,53.67,113.015,208.0525,518.55
Price,73100.0,NaN,NaN,NaN,55.135108,26.021945,10.0,32.65,55.05,77.86,100.0


In [6]:
df["Store ID"].unique() # What Stores exist?

array(['S001', 'S002', 'S003', 'S004', 'S005'], dtype=object)

In [7]:
df["Product ID"].unique() # What products exist?

array(['P0001', 'P0002', 'P0003', 'P0004', 'P0005', 'P0006', 'P0007',
       'P0008', 'P0009', 'P0010', 'P0011', 'P0012', 'P0013', 'P0014',
       'P0015', 'P0016', 'P0017', 'P0018', 'P0019', 'P0020'], dtype=object)

In [8]:
df["Category"].unique() # What categories exist?

array(['Groceries', 'Toys', 'Electronics', 'Furniture', 'Clothing'],
      dtype=object)

In [9]:
df["Region"].unique() # What are the different regions?

array(['North', 'South', 'West', 'East'], dtype=object)

In [10]:
df["Date"].min(), df["Date"].max() # What is the date range?

('2022-01-01', '2024-01-01')

In [11]:
df.isna().sum() # Any missing values?

Date                  0
Store ID              0
Product ID            0
Category              0
Region                0
Inventory Level       0
Units Sold            0
Units Ordered         0
Demand Forecast       0
Price                 0
Discount              0
Weather Condition     0
Holiday/Promotion     0
Competitor Pricing    0
Seasonality           0
dtype: int64

In [12]:
df.duplicated().sum() # Any duplicate rows?

0

In [13]:
# Verify the number of daily observations per Store × Product combination

# and how many combinations have each observation count.

# Output: 731 observations     100 Store × Product combinations

df.groupby(["Store ID", "Product ID"]).size().value_counts().sort_index()

731    100
Name: count, dtype: int64

In [14]:
# Verify the first few dates represented in the dataset

df["Date"].value_counts().sort_index().head()

Date
2022-01-01    100
2022-01-02    100
2022-01-03    100
2022-01-04    100
2022-01-05    100
Name: count, dtype: int64

In [15]:
# Verify the last few dates represented in the dataset

df["Date"].value_counts().sort_index().tail()

Date
2023-12-28    100
2023-12-29    100
2023-12-30    100
2023-12-31    100
2024-01-01    100
Name: count, dtype: int64

In [16]:
# Verify that every date has the expected number of Store × Product observations

df.groupby("Date").size().value_counts().sort_index()

100    731
Name: count, dtype: int64

In [17]:
# Check the number of unique values in every column

df.nunique()

Date                    731
Store ID                  5
Product ID               20
Category                  5
Region                    4
Inventory Level         451
Units Sold              498
Units Ordered           181
Demand Forecast       31608
Price                  8999
Discount                  5
Weather Condition         4
Holiday/Promotion         2
Competitor Pricing     9751
Seasonality               4
dtype: int64

In [18]:
# Check the data type assigned to each column to ensure Pandas interpreted the dataset correctly

df.dtypes

Date                   object
Store ID               object
Product ID             object
Category               object
Region                 object
Inventory Level         int64
Units Sold              int64
Units Ordered           int64
Demand Forecast       float64
Price                 float64
Discount                int64
Weather Condition      object
Holiday/Promotion       int64
Competitor Pricing    float64
Seasonality            object
dtype: object

In [19]:
# Convert the Date column from text (object) into Pandas' datetime type for proper date-based analysis

df["Date"] = pd.to_datetime(df["Date"])

In [20]:
# Verify that the Date column was successfully converted to datetime

df["Date"].dtype

dtype('<M8[ns]')

In [21]:
# Confirm that the dataset still has 731 unique dates after converting Date to datetime

df["Date"].nunique()

731

In [22]:
# Verify that each Store × Product combination has one observation for every date in the dataset

store_product_date_counts = df.groupby(["Store ID", "Product ID"])["Date"].nunique()

store_product_date_counts.value_counts().sort_index()

Date
731    100
Name: count, dtype: int64

In [23]:
# Display all columns and their Pandas data types as a quick data dictionary reference

data_dictionary = pd.DataFrame({
    "Column": df.columns,
    "Data Type": df.dtypes.astype(str).values
})

data_dictionary

,Column,Data Type
0,Date,datetime64[ns]
1,Store ID,object
2,Product ID,object
3,Category,object
4,Region,object
5,Inventory Level,int64
6,Units Sold,int64
7,Units Ordered,int64
8,Demand Forecast,float64
9,Price,float64


In [24]:
# Summarize the minimum, maximum, mean, and median of each numerical column to identify suspicious values

df.describe().T[["min", "max", "mean", "50%"]]

,min,max,mean,50%
Date,2022-01-01 00:00:00,2024-01-01 00:00:00,2022-12-31 23:59:59.999999744,2023-01-01 00:00:00
Inventory Level,50.0,500.0,274.469877,273.0
Units Sold,0.0,499.0,136.46487,107.0
Units Ordered,20.0,200.0,110.004473,110.0
Demand Forecast,-9.99,518.55,141.49472,113.015
Price,10.0,100.0,55.135108,55.05
Discount,0.0,20.0,10.009508,10.0
Holiday/Promotion,0.0,1.0,0.497305,0.0
Competitor Pricing,5.03,104.94,55.146077,55.01


### Dataset Quality & Structure

The dataset is clean and structurally consistent, with 73,100 rows across 5 stores, 20 products, and 731 days. 

No missing values or duplicates were found. 

Most numerical ranges look reasonable, but negative demand forecasts require further investigation.

In [25]:
# Find all rows where the forecasted demand is negative so we can investigate the anomaly

negative_forecasts = df[df["Demand Forecast"] < 0]

negative_forecasts

,Date,Store ID,Product ID,Category,Region,Inventory Level,Units Sold,Units Ordered,Demand Forecast,Price,Discount,Weather Condition,Holiday/Promotion,Competitor Pricing,Seasonality
63,2022-01-01,S004,P0004,Groceries,West,437,0,160,-2.40,87.23,10,Sunny,0,90.36,Spring
141,2022-01-02,S003,P0002,Groceries,East,175,2,140,-3.40,87.50,0,Rainy,0,91.20,Autumn
278,2022-01-03,S004,P0019,Toys,North,140,1,47,-3.91,80.41,0,Snowy,0,76.90,Winter
511,2022-01-06,S001,P0012,Groceries,North,59,1,88,-8.37,63.21,0,Rainy,0,60.48,Spring
730,2022-01-08,S002,P0011,Electronics,South,64,7,79,-2.99,92.43,0,Snowy,0,88.55,Spring
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
72716,2023-12-29,S001,P0017,Clothing,North,487,7,132,-0.73,55.34,0,Sunny,1,58.03,Summer
72722,2023-12-29,S002,P0003,Toys,West,199,4,180,-5.50,82.08,10,Cloudy,1,77.99,Spring
72859,2023-12-30,S003,P0020,Toys,North,203,1,119,-7.68,94.33,5,Sunny,1,96.79,Autumn
73026,2024-01-01,S002,P0007,Toys,West,53,2,40,-6.08,63.66,20,Snowy,1,62.37,Autumn


In [26]:
# Count how many negative demand forecasts exist and what percentage of the dataset they represent
print("Negative forecasts:", len(negative_forecasts))
print("Percentage of dataset:", len(negative_forecasts) / len(df) * 100)

Negative forecasts: 673
Percentage of dataset: 0.9206566347469219


### Anomaly Investigation Logic

If negative forecasts are spread fairly evenly across the dataset, they are likely generated-data noise. If they are heavily concentrated in a specific product, store, or season, that may indicate a meaningful pattern worth investigating.

In [27]:
# Count negative demand forecasts for each product to see whether the anomaly is concentrated in specific products

negative_forecasts["Product ID"].value_counts()

Product ID
P0019    50
P0015    49
P0010    38
P0004    37
P0003    37
P0007    37
P0001    37
P0020    35
P0011    34
P0008    33
P0002    33
P0009    32
P0017    32
P0013    32
P0012    32
P0014    30
P0018    28
P0005    27
P0006    24
P0016    16
Name: count, dtype: int64

In [28]:
# Count negative demand forecasts for each store to see whether the anomaly is concentrated in specific stores

negative_forecasts["Store ID"].value_counts()

Store ID
S003    143
S001    141
S002    137
S005    132
S004    120
Name: count, dtype: int64

In [29]:
# Count negative demand forecasts by season to see whether the anomaly is related to seasonal patterns

negative_forecasts["Seasonality"].value_counts()

Seasonality
Spring    192
Summer    172
Winter    156
Autumn    153
Name: count, dtype: int64

### Anomaly Investigation Conclusion

The negative forecasts appear to be distributed across the dataset rather than concentrated in one product, store, or season.

This suggests synthetic-data noise or anomaly generation rather than a meaningful business pattern.

I will document the finding, retain the values for now, and move on.

In [30]:
# Measure the linear relationship between inventory and sales to see whether higher inventory generally corresponds to higher sales

df[["Inventory Level", "Units Sold"]].corr()

,Inventory Level,Units Sold
Inventory Level,1.000000,0.589995
Units Sold,0.589995,1.000000


For reference, 

The correlation ranges from -1 to +1:

+1 → strong positive relationship

0 → little/no linear relationship

-1 → strong negative relationship

In [31]:
# Calculate the average absolute difference between forecasted demand and actual units sold

# to get an initial measure of how far the forecasts are from observed sales.

forecast_error = (df["Demand Forecast"] - df["Units Sold"]).abs().mean()

print("Mean Absolute Forecast Error:", forecast_error)

Mean Absolute Forecast Error: 8.338075512995895


An MAE of 8.34 establishes that the forecasts are reasonably close to actual sales on average.

In [32]:
# Calculate forecast error for every row so we can distinguish over-forecasting from under-forecasting

df["Forecast Error"] = df["Demand Forecast"] - df["Units Sold"]

# Display summary statistics of the forecast error

df["Forecast Error"].describe()

count    73100.000000
mean         5.029850
std          8.660766
min        -10.000000
25%         -2.470000
50%          4.990000
75%         12.510000
max         20.000000
Name: Forecast Error, dtype: float64

The forecast error has a mean of +5.03 units, indicating a consistent tendency to over-forecast demand by approximately 5 units. 

The similar median error of +4.99 suggests that this bias is widespread rather than driven only by extreme observations.

In [33]:
# Measure how strongly forecasted demand and actual sales move together

df[["Demand Forecast", "Units Sold"]].corr()

,Demand Forecast,Units Sold
Demand Forecast,1.000000,0.996853
Units Sold,0.996853,1.000000


### Forecast vs Actual Demand

Forecast and Units Sold have an extremely strong positive correlation of 0.997, indicating that the forecasts closely track actual demand. 

Combined with the +5.03 average forecast error, the dataset shows highly aligned forecasts with a slight over-forecasting bias.

In [34]:
# Calculate key inventory percentiles to understand what "low" and "high" inventory look like in this dataset

df["Inventory Level"].quantile([0.10, 0.25, 0.50, 0.75, 0.90])

0.10     95.0
0.25    162.0
0.50    273.0
0.75    387.0
0.90    455.0
Name: Inventory Level, dtype: float64

In [35]:
# Define the 10th percentile as a temporary threshold for identifying unusually low inventory

low_inventory_threshold = df["Inventory Level"].quantile(0.10)

low_inventory_threshold

95.0

In [36]:
# Count how many observations fall into the bottom 10% of inventory levels

low_inventory_count = (df["Inventory Level"] <= low_inventory_threshold).sum()

print("Low inventory observations:", low_inventory_count)

print("Percentage of dataset:", low_inventory_count / len(df) * 100)

Low inventory observations: 7387
Percentage of dataset: 10.105335157318741


In [37]:
# Count low-inventory observations for each store to identify stores that may experience inventory pressure more frequently

df[df["Inventory Level"] <= low_inventory_threshold]["Store ID"].value_counts()

Store ID
S002    1561
S004    1488
S001    1486
S005    1452
S003    1400
Name: count, dtype: int64

In [38]:
# Calculate the percentage of observations with low inventory for each store
# so stores can be compared using a normalized metric rather than raw counts.

low_inventory_by_store = (
    df.assign(Low_Inventory=df["Inventory Level"] <= low_inventory_threshold)
      .groupby("Store ID")["Low_Inventory"]
      .mean()
      .mul(100)
      .sort_values(ascending=False)
)

low_inventory_by_store

Store ID
S002    10.677155
S004    10.177839
S001    10.164159
S005     9.931601
S003     9.575923
Name: Low_Inventory, dtype: float64

### Store-Level Inventory Analysis
Low-inventory observations are relatively evenly distributed across stores, ranging from 9.58% to 10.68%. S002 has the highest rate, while S003 has the lowest, but the small difference suggests no significant store-level imbalance.

In [39]:
# Calculate the percentage of observations with low inventory for each product
# to identify products that experience inventory pressure more frequently.

low_inventory_by_product = (
    df.assign(Low_Inventory=df["Inventory Level"] <= low_inventory_threshold)

      .groupby("Product ID")["Low_Inventory"]

      .mean()

      .mul(100)

      .sort_values(ascending=False)
)

low_inventory_by_product

Product ID
P0006    10.971272
P0011    10.670315
P0004    10.642955
P0002    10.560876
P0016    10.396717
P0007    10.232558
P0003    10.205198
P0012    10.205198
P0001    10.150479
P0013    10.123119
P0008    10.041040
P0005    10.041040
P0015    10.013680
P0009     9.986320
P0019     9.958960
P0020     9.822161
P0010     9.794802
P0018     9.767442
P0017     9.493844
P0014     9.028728
Name: Low_Inventory, dtype: float64

### Product-Level Inventory Analysis
Low-inventory observations are relatively evenly distributed across products, ranging from 9.29% to 10.97%. P0006 has the highest rate and P0014 the lowest, but the small difference suggests no significant product-level imbalance.

In [40]:
# Calculate an approximate inventory coverage ratio to estimate how many days
# the current inventory could cover at the observed daily sales rate.

df["Inventory Coverage"] = df["Inventory Level"] / df["Units Sold"].replace(0, np.nan) # Replace zero sales values with NaN so they don't create invalid division results

df["Inventory Coverage"].describe()

count    72740.000000
mean         6.165394
std         21.061941
min          1.000000
25%          1.333333
50%          2.007407
75%          4.000000
max        493.000000
Name: Inventory Coverage, dtype: float64

In [41]:
# Summarize the distribution of inventory coverage to understand typical and extreme coverage levels.

df["Inventory Coverage"].quantile([0.10, 0.25, 0.50, 0.75, 0.90])

0.10    1.107914
0.25    1.333333
0.50    2.007407
0.75    4.000000
0.90    9.782609
Name: Inventory Coverage, dtype: float64

In [42]:
# Calculate inventory coverage using forecasted demand as the expected daily demand rate
# so coverage reflects anticipated demand rather than only today's sales.

df["Forecast Coverage"] = (
    df["Inventory Level"] / df["Demand Forecast"].replace(0, np.nan)
)

In [43]:
# Examine the distribution of forecast-based inventory coverage to understand expected stock duration

df["Forecast Coverage"].quantile([0.10, 0.25, 0.50, 0.75, 0.90])

0.10    1.067745
0.25    1.275534
0.50    1.887534
0.75    3.537469
0.90    7.698941
Name: Forecast Coverage, dtype: float64

### Forecast-Based Inventory Coverage

The median inventory coverage is about 1.89 days.

25% of the observations have less than 1.28 days of coverage, while 75% have less than 3.57 days.

This means a large portion of the data has relatively low inventory coverage based on forecasted demand.

In [44]:
# Identify observations where available inventory is less than one day of forecasted demand

low_coverage = df["Forecast Coverage"] < 1

print("Low coverage observations:", low_coverage.sum())

print("Percentage of dataset:", low_coverage.mean() * 100)

Low coverage observations: 3258
Percentage of dataset: 4.456908344733242


### Low Inventory Coverage

3,258 observations (4.46%) have less than 1 day of inventory coverage based on forecasted demand.

These are potential inventory-risk situations that we can investigate later with ShelfSleuth.

-----------------------------------------------------------------------------------------------------------------------------------------------------

### Move Data to DuckDB

We have finished the basic data exploration.

Now we'll put our DataFrame into a SQL database so we can start working with SQL and eventually build our Text-to-SQL and semantic layer components.

In [45]:
# Import DuckDB and Path so we can create and connect to our local database

import duckdb

from pathlib import Path


# Create the folder where our DuckDB database will be stored

db_dir = Path("../data/processed")

db_dir.mkdir(parents=True, exist_ok=True)


# Create or open the ShelfSleuth DuckDB database

db_path = db_dir / "shelfsleuth.duckdb"

con = duckdb.connect(str(db_path))

In [46]:
# Register our Pandas DataFrame with DuckDB so DuckDB can access it

con.register("retail_df", df)


# Copy the DataFrame into a persistent SQL table called retail_inventory

con.execute("""
    CREATE OR REPLACE TABLE retail_inventory AS
    SELECT * FROM retail_df
""")

In [47]:
# Check that our SQL table contains all 73,100 rows from the original DataFrame

con.execute("SELECT COUNT(*) FROM retail_inventory").fetchone()

(73100,)

### SQL Exploration

Our data is now stored as a SQL table in DuckDB.

We'll use a few SQL queries to understand how the table works before building the semantic layer on top of it.

In [48]:
# Show the SQL schema so we can see the columns and data types available to the Text-to-SQL agent

con.execute("DESCRIBE retail_inventory").fetchdf()

,column_name,column_type,null,key,default,extra
0,Date,TIMESTAMP_NS,YES,None,None,None
1,Store ID,VARCHAR,YES,None,None,None
2,Product ID,VARCHAR,YES,None,None,None
3,Category,VARCHAR,YES,None,None,None
4,Region,VARCHAR,YES,None,None,None
5,Inventory Level,BIGINT,YES,None,None,None
6,Units Sold,BIGINT,YES,None,None,None
7,Units Ordered,BIGINT,YES,None,None,None
8,Demand Forecast,DOUBLE,YES,None,None,None
9,Price,DOUBLE,YES,None,None,None


In [49]:
# Find the average inventory level for each store using SQL

con.execute("""
    SELECT
        "Store ID",
        AVG("Inventory Level") AS avg_inventory

    FROM retail_inventory

    GROUP BY "Store ID"

    ORDER BY avg_inventory DESC
""").fetchdf()

,Store ID,avg_inventory
0,S003,276.872298
1,S005,276.504446
2,S001,273.339945
3,S002,273.327907
4,S004,272.304788


### Semantic Layer

The database tells us what the columns are, but not what they mean from a business perspective.

The semantic layer will define things like metrics, dimensions, and business terms so the AI knows how to translate user questions into the right SQL.

In [50]:
# Add the ShelfSleuth project root to Python's import path so the notebook can access our project modules

import sys

from pathlib import Path


# Move one level up from the notebooks directory to reach the ShelfSleuth project root

project_root = Path.cwd().parent # cwd --> current working directory

sys.path.append(str(project_root))

In [51]:
# Import the semantic layer so we can inspect the business definitions available to the AI.

from semantic_layer.metrics import get_semantic_context # get_semantic_context is a function in ShelfSleuth//semantic_layer/metrics.py


# Retrieve the complete semantic context.

semantic_context = get_semantic_context()


# Display the semantic definitions.

semantic_context

{'metrics': {'average_inventory': {'description': 'Average inventory level across observations',
   'expression': 'AVG("Inventory Level")'},
  'units_sold': {'description': 'Total number of units sold',
   'expression': 'SUM("Units Sold")'},
  'demand_forecast': {'description': 'Total forecasted demand',
   'expression': 'SUM("Demand Forecast")'},
  'inventory_coverage': {'description': 'Inventory available relative to forecasted demand',
   'expression': 'AVG("Inventory Level" / NULLIF("Demand Forecast", 0))'}},
 'dimensions': {'store': '"Store ID"',
  'product': '"Product ID"',
  'category': '"Category"',
  'region': '"Region"',
  'date': '"Date"',
  'seasonality': '"Seasonality"'},
 'business_rules': {'low_inventory_coverage': {'description': 'Inventory is less than one day of forecasted demand',
   'condition': '"Inventory Level" < "Demand Forecast"'}}}

In [52]:
# Import the Text-to-SQL agent so we can connect it to DuckDB and our semantic layer

from agents.text_to_sql_agent import TextToSQLAgent


# Create the Text-to-SQL agent

text_to_sql_agent = TextToSQLAgent(
    database_connection=con,

    semantic_context=semantic_context,
)


# Display the agent to confirm it was created successfully

text_to_sql_agent

In [53]:
# Test whether our Text-to-SQL agent can execute a SQL query against DuckDB

test_sql = """
SELECT
    "Store ID",
    AVG("Inventory Level") AS avg_inventory
FROM retail_inventory
GROUP BY "Store ID"
ORDER BY avg_inventory DESC
"""


# Execute the SQL through our agent

result = text_to_sql_agent.execute_sql(test_sql)


# Display the result

result

,Store ID,avg_inventory
0,S003,276.872298
1,S005,276.504446
2,S001,273.339945
3,S002,273.327907
4,S004,272.304788


In [54]:
# Ask the Text-to-SQL agent to convert a natural-language question into SQL

question = "Which store has the lowest average inventory?"


# Generate SQL using Gemini

generated_sql = text_to_sql_agent.generate_sql(question)


# Display the SQL generated by Gemini

print(generated_sql)

Direct use of automatic function calling (AFC) in Models.generate_content is not recommended. Instead, we recommend to use AFC in Chat.send_message. Similarly, direct use of AFC in Models.generate_content_stream is not recommended. Instead, we recommend to use AFC in Chat.send_message_stream.


SELECT "Store ID", AVG("Inventory Level") AS "average_inventory"
FROM retail_inventory
GROUP BY "Store ID"
ORDER BY AVG("Inventory Level") ASC
LIMIT 1


In [55]:
# Execute the SQL generated by Gemini against our DuckDB database

result = text_to_sql_agent.execute_sql(generated_sql)


# Display the final result returned by the database

result

,Store ID,average_inventory
0,S004,272.304788


In [56]:
# Ask the agent to answer a business question using its complete SQL generation and execution loop

question = "Which store has the lowest average inventory?"


# Let the agent generate SQL, execute it, and automatically correct SQL errors if necessary

answer = text_to_sql_agent.answer_question(question)


# Display the SQL generated by the agent

print(answer["sql"])


# Display the final database result

answer["result"]

SELECT "Store ID", AVG("Inventory Level") AS "average_inventory"
FROM retail_inventory
GROUP BY "Store ID"
ORDER BY "average_inventory" ASC
LIMIT 1;


,Store ID,average_inventory
0,S004,272.304788


### Text-to-SQL Agent Loop

The agent now does the same process as our previous generate_sql() → execute_sql() snippets, but handles everything through one answer_question() method.

It can also detect SQL errors, ask Gemini to correct them, and retry.

This is our first real agent loop.

Next: OKF/Knowledge Agent — combining SQL results with business knowledge to explain why something is happening.

In [57]:
# Import the Knowledge Agent and our business knowledge base

from agents.knowledge_agent import KnowledgeAgent

from okf.knowledge_base import KNOWLEDGE_BASE


# Create the Knowledge Agent

knowledge_agent = KnowledgeAgent(
    knowledge_base=KNOWLEDGE_BASE
)

In [58]:
from tools.investigation_tool import InvestigationTool

investigation_tool = InvestigationTool(
    database_connection=con
)

print("Investigation tool initialized successfully.")

Investigation tool initialized successfully.


In [62]:
import importlib

import agents.action_planner_agent
import agents.critic_agent
import agents.research_pipeline

importlib.reload(agents.action_planner_agent)
importlib.reload(agents.critic_agent)
importlib.reload(agents.research_pipeline)

print("ShelfSleuth agents reloaded successfully.")

ShelfSleuth agents reloaded successfully.


In [63]:
from agents.research_pipeline import ShelfSleuthPipeline

pipeline = ShelfSleuthPipeline(
    database_connection=con,
    semantic_context=semantic_context,
    knowledge_base=KNOWLEDGE_BASE,
    investigation_tool=investigation_tool,
)

print("ShelfSleuth pipeline initialized successfully.")

ShelfSleuth pipeline initialized successfully.


In [64]:
answer = pipeline.run(question)

print("=" * 80)
print("SHELF SLEUTH INVESTIGATION")
print("=" * 80)

print("\nQUESTION")
print(answer["question"])

print("\nSQL")
print(answer["sql"])

print("\nSQL RESULT")
print(answer["sql_result"])

print("\nBUSINESS KNOWLEDGE")
print(answer["knowledge"])

print("\nROOT CAUSE ANALYSIS")
print(answer["root_cause_analysis"])

print("\nACTION PLAN")
print(answer["action_plan"])

print("\nCRITIC REVIEW")
print(answer["critic_review"])

SHELF SLEUTH INVESTIGATION

QUESTION
Which store has the lowest average inventory?

SQL
SELECT
  "Store ID",
  AVG("Inventory Level") AS "average_inventory"
FROM retail_inventory
GROUP BY
  "Store ID"
ORDER BY
  "average_inventory" ASC
LIMIT 1;

SQL RESULT
  Store ID  average_inventory
0     S004         272.304788

BUSINESS KNOWLEDGE
{'inventory': ['Low inventory can increase the risk of stockouts.', 'Inventory should be evaluated relative to expected demand, not only as an absolute quantity.', 'Inventory coverage below one day indicates a potential inventory-risk situation.'], 'demand': ['Demand Forecast represents expected demand for the observation.', 'A persistent gap between forecasted demand and actual units sold may indicate forecast bias.'], 'store_performance': ['Store-level inventory differences should be investigated alongside sales and demand patterns.', 'A store with low inventory is not necessarily underperforming if its demand is also low.']}

ROOT CAUSE ANALYSIS
{'inve

In [65]:
# ShelfSleuth robustness test questions

test_questions = [
    "Which store has the highest average units sold?",
    "Which store has the largest gap between demand forecast and actual units sold?",
    "Which store has the lowest inventory coverage?",
    "Which store has the highest average inventory?",
    "Which store appears to have the greatest inventory risk?"
]

print("ShelfSleuth robustness test set created.")
print(f"Number of test questions: {len(test_questions)}")

ShelfSleuth robustness test set created.
Number of test questions: 5


In [66]:
# Establish deterministic ground truth for robustness testing

ground_truth = {}

ground_truth["highest_average_units_sold"] = con.execute("""
    SELECT
        "Store ID",
        AVG("Units Sold") AS average_units_sold
    FROM retail_inventory
    GROUP BY "Store ID"
    ORDER BY average_units_sold DESC
    LIMIT 1
""").fetchdf()

ground_truth["largest_forecast_sales_gap"] = con.execute("""
    SELECT
        "Store ID",
        AVG("Demand Forecast" - "Units Sold") AS forecast_sales_gap
    FROM retail_inventory
    GROUP BY "Store ID"
    ORDER BY forecast_sales_gap DESC
    LIMIT 1
""").fetchdf()

ground_truth["lowest_inventory_coverage"] = con.execute("""
    SELECT
        "Store ID",
        AVG("Inventory Level" / NULLIF("Demand Forecast", 0)) AS inventory_coverage
    FROM retail_inventory
    GROUP BY "Store ID"
    ORDER BY inventory_coverage ASC
    LIMIT 1
""").fetchdf()

ground_truth["highest_average_inventory"] = con.execute("""
    SELECT
        "Store ID",
        AVG("Inventory Level") AS average_inventory
    FROM retail_inventory
    GROUP BY "Store ID"
    ORDER BY average_inventory DESC
    LIMIT 1
""").fetchdf()

# For "greatest inventory risk", use the lowest average inventory coverage
ground_truth["greatest_inventory_risk"] = con.execute("""
    SELECT
        "Store ID",
        AVG("Inventory Level" / NULLIF("Demand Forecast", 0)) AS inventory_coverage
    FROM retail_inventory
    GROUP BY "Store ID"
    ORDER BY inventory_coverage ASC
    LIMIT 1
""").fetchdf()


print("=" * 80)
print("SHELF SLEUTH — GROUND TRUTH")
print("=" * 80)

for test_name, result in ground_truth.items():
    print(f"\n{test_name}")
    print(result)

SHELF SLEUTH — GROUND TRUTH

highest_average_units_sold
  Store ID  average_units_sold
0     S003            138.3513

largest_forecast_sales_gap
  Store ID  forecast_sales_gap
0     S005              5.1252

lowest_inventory_coverage
  Store ID  inventory_coverage
0     S002            2.913299

highest_average_inventory
  Store ID  average_inventory
0     S003         276.872298

greatest_inventory_risk
  Store ID  inventory_coverage
0     S002            2.913299


In [68]:
# Build a clean evaluation table from the deterministic ground truth

import pandas as pd

evaluation_ground_truth = []

for test_name, result in ground_truth.items():

    row = result.iloc[0]

    evaluation_ground_truth.append({
        "test": test_name,
        "store_id": row["Store ID"],
        "metric_value": list(row)[1],
    })

evaluation_df = pd.DataFrame(evaluation_ground_truth)

print("=" * 80)
print("SHELF SLEUTH — EVALUATION GROUND TRUTH")
print("=" * 80)

display(evaluation_df)

SHELF SLEUTH — EVALUATION GROUND TRUTH


,test,store_id,metric_value
0,highest_average_units_sold,S003,138.351300
1,largest_forecast_sales_gap,S005,5.125200
2,lowest_inventory_coverage,S002,2.913299
3,highest_average_inventory,S003,276.872298
4,greatest_inventory_risk,S002,2.913299


In [71]:
# Complete the 5-question ShelfSleuth evaluation
#
# IMPORTANT:
# - Reuses successful tests already stored in shelfsleuth_results
# - Only runs questions that have NOT completed successfully
# - This is the ONE intentional Gemini-heavy evaluation run
# - DO NOT rerun this cell after it finishes

print("=" * 80)
print("SHELF SLEUTH — 5 QUESTION EVALUATION")
print("=" * 80)

# Keep any successful results we already have
existing_successful_questions = {
    result["question"]
    for result in shelfsleuth_results
    if result["status"] == "SUCCESS"
}

print(
    f"\nAlready completed successfully: "
    f"{len(existing_successful_questions)}/{len(test_questions)}"
)

remaining_questions = [
    q for q in test_questions
    if q not in existing_successful_questions
]

print(
    f"Remaining pipeline runs: "
    f"{len(remaining_questions)}"
)

for i, test_question in enumerate(remaining_questions, start=1):

    print("\n" + "=" * 80)
    print(f"RUNNING REMAINING TEST {i}/{len(remaining_questions)}")
    print("=" * 80)
    print(f"Question: {test_question}")

    try:

        result = pipeline.run(test_question)

        shelfsleuth_results.append({
            "question": test_question,
            "result": result,
            "status": "SUCCESS",
        })

        print("\nStatus: SUCCESS")
        print("\nSQL:")
        print(result["sql"])

        print("\nSQL Result:")
        print(result["sql_result"])

    except Exception as error:

        shelfsleuth_results.append({
            "question": test_question,
            "result": None,
            "status": "FAILED",
            "error": str(error),
        })

        print("\nStatus: FAILED")
        print(f"Error: {error}")


print("\n" + "=" * 80)
print("5-QUESTION EVALUATION RUN COMPLETE")
print("=" * 80)

successful = sum(
    1
    for result in shelfsleuth_results
    if result["status"] == "SUCCESS"
)

failed = sum(
    1
    for result in shelfsleuth_results
    if result["status"] == "FAILED"
)

print(f"Successful tests: {successful}/{len(test_questions)}")
print(f"Failed tests: {failed}/{len(test_questions)}")

SHELF SLEUTH — 5 QUESTION EVALUATION

Already completed successfully: 1/5
Remaining pipeline runs: 4

RUNNING REMAINING TEST 1/4
Question: Which store has the largest gap between demand forecast and actual units sold?

Status: FAILED
Error: Gemini API quota has been exhausted. Please wait for the quota to reset.

RUNNING REMAINING TEST 2/4
Question: Which store has the lowest inventory coverage?

Status: FAILED
Error: 429 RESOURCE_EXHAUSTED. {'error': {'code': 429, 'message': 'You exceeded your current quota, please check your plan and billing details. For more information on this error, head to: https://ai.google.dev/gemini-api/docs/rate-limits. To monitor your current usage, head to: https://ai.dev/rate-limit. \n* Quota exceeded for metric: generativelanguage.googleapis.com/generate_content_free_tier_requests, limit: 20, model: gemini-3.6-flash\nPlease retry in 15.946215954s.', 'status': 'RESOURCE_EXHAUSTED', 'details': [{'@type': 'type.googleapis.com/google.rpc.Help', 'links': [{'de

In [72]:
# Inspect the results from the live ShelfSleuth tests
# This uses results already generated and makes NO new Gemini calls.

print("=" * 80)
print("SHELF SLEUTH — COMPLETED LIVE TESTS")
print("=" * 80)

completed_tests = []

for i, test_result in enumerate(shelfsleuth_results, start=1):

    if test_result["status"] != "SUCCESS":
        continue

    question = test_result["question"]
    result = test_result["result"]

    print(f"\n{'-' * 80}")
    print(f"TEST {i}")
    print(f"{'-' * 80}")

    print("\nQUESTION")
    print(question)

    print("\nGENERATED SQL")
    print(result["sql"])

    print("\nSQL RESULT")
    print(result["sql_result"])

    print("\nROOT CAUSE ANALYSIS")
    print(result["root_cause_analysis"])

    print("\nACTION PLAN")
    print(result["action_plan"])

    print("\nCRITIC REVIEW")
    print(result["critic_review"])

    completed_tests.append(test_result)


print("\n" + "=" * 80)
print(f"COMPLETED SUCCESSFUL TESTS: {len(completed_tests)}")
print("=" * 80)

SHELF SLEUTH — COMPLETED LIVE TESTS

--------------------------------------------------------------------------------
TEST 1
--------------------------------------------------------------------------------

QUESTION
Which store has the highest average units sold?

GENERATED SQL
SELECT "Store ID"
FROM retail_inventory
GROUP BY "Store ID"
ORDER BY AVG("Units Sold") DESC
LIMIT 1;

SQL RESULT
  Store ID
0     S003

ROOT CAUSE ANALYSIS
{'investigation': 'store_inventory_comparison', 'investigation_result':   Store ID  avg_inventory  avg_units_sold  avg_demand_forecast  \
0     S004     272.304788      135.379275           140.400666   
1     S002     273.327907      135.958618           140.973995   
2     S001     273.339945      135.140219           140.144895   
3     S005     276.504446      137.494938           142.620138   
4     S003     276.872298      138.351300           143.333906   

   avg_inventory_coverage  
0                3.852037  
1                2.913299  
2           

### **Robustness & Evaluation**

Instead of repeatedly calling Gemini, we will use the completed runs to evaluate whether:

- the generated SQL answers the business question correctly
- the SQL result matches the deterministic ground truth
- the Root Cause Agent interprets the data appropriately
- the Action Planner produces evidence-based recommendations
- the Critic Agent identifies unsupported conclusions

This gives us a basic evaluation framework while keeping Gemini usage controlled.

**Next: Evaluation Framework — comparing ShelfSleuth outputs against deterministic ground truth and identifying where the pipeline can be improved.**

In [73]:
# Compare completed ShelfSleuth SQL results against deterministic ground truth

evaluation_results = []

for test_result in completed_tests:

    question = test_result["question"]
    result = test_result["result"]

    # Extract the Store ID returned by ShelfSleuth
    sql_result = result["sql_result"]

    predicted_store = str(sql_result.iloc[0]["Store ID"])

    # Match the question to the appropriate ground-truth test
    if "highest average units sold" in question.lower():
        test_name = "highest_average_units_sold"

    elif "largest gap between demand forecast and actual units sold" in question.lower():
        test_name = "largest_forecast_sales_gap"

    elif "lowest inventory coverage" in question.lower():
        test_name = "lowest_inventory_coverage"

    elif "highest average inventory" in question.lower():
        test_name = "highest_average_inventory"

    elif "greatest inventory risk" in question.lower():
        test_name = "greatest_inventory_risk"

    else:
        test_name = None

    # Compare ShelfSleuth's answer with the deterministic answer
    if test_name is not None:

        expected_store = str(
            ground_truth[test_name].iloc[0]["Store ID"]
        )

        passed = predicted_store == expected_store

        evaluation_results.append({
            "test": test_name,
            "predicted_store": predicted_store,
            "expected_store": expected_store,
            "status": "PASS" if passed else "FAIL",
        })


evaluation_results_df = pd.DataFrame(evaluation_results)

print("=" * 80)
print("SHELF SLEUTH — LIVE TEST EVALUATION")
print("=" * 80)

display(evaluation_results_df)

if len(evaluation_results_df) > 0:
    accuracy = (
        evaluation_results_df["status"].eq("PASS").mean() * 100
    )

    print(f"\nEvaluation accuracy: {accuracy:.1f}%")
    print(
        f"Tests evaluated: "
        f"{len(evaluation_results_df)}"
    )

SHELF SLEUTH — LIVE TEST EVALUATION


,test,predicted_store,expected_store,status
0,highest_average_units_sold,S003,S003,PASS



Evaluation accuracy: 100.0%
Tests evaluated: 1


In [74]:
# Clean evaluation results
# No Gemini calls — this only removes duplicate historical test records.

unique_results = {}

for test_result in shelfsleuth_results:

    question = test_result["question"]

    # Keep the successful result if one exists.
    # Otherwise keep the most recent result.
    if question not in unique_results:
        unique_results[question] = test_result

    elif test_result["status"] == "SUCCESS":
        unique_results[question] = test_result


clean_shelfsleuth_results = list(unique_results.values())

print("=" * 80)
print("SHELF SLEUTH — CLEAN TEST RESULTS")
print("=" * 80)

for i, test_result in enumerate(
    clean_shelfsleuth_results,
    start=1
):
    print(
        f"{i}. "
        f"{test_result['status']} — "
        f"{test_result['question']}"
    )

print(
    f"\nUnique questions recorded: "
    f"{len(clean_shelfsleuth_results)}"
)

print(
    "Successful live tests: "
    f"{sum(r['status'] == 'SUCCESS' for r in clean_shelfsleuth_results)}"
)

print(
    "Failed due to quota/errors: "
    f"{sum(r['status'] == 'FAILED' for r in clean_shelfsleuth_results)}"
)

SHELF SLEUTH — CLEAN TEST RESULTS
1. SUCCESS — Which store has the highest average units sold?
2. FAILED — Which store has the largest gap between demand forecast and actual units sold?
3. FAILED — Which store has the lowest inventory coverage?
4. FAILED — Which store has the highest average inventory?
5. FAILED — Which store appears to have the greatest inventory risk?

Unique questions recorded: 5
Successful live tests: 1
Failed due to quota/errors: 4


In [75]:
# Summarize the ShelfSleuth evaluation
# No Gemini calls.

successful_tests = [
    result
    for result in clean_shelfsleuth_results
    if result["status"] == "SUCCESS"
]

failed_tests = [
    result
    for result in clean_shelfsleuth_results
    if result["status"] == "FAILED"
]

evaluation_summary = pd.DataFrame([
    {
        "metric": "Total test questions",
        "value": len(test_questions),
    },
    {
        "metric": "Successful live tests",
        "value": len(successful_tests),
    },
    {
        "metric": "Quota/error blocked tests",
        "value": len(failed_tests),
    },
    {
        "metric": "Live tests passed",
        "value": sum(
            result["status"] == "SUCCESS"
            for result in successful_tests
        ),
    },
])

print("=" * 80)
print("SHELF SLEUTH — EVALUATION SUMMARY")
print("=" * 80)

display(evaluation_summary)

SHELF SLEUTH — EVALUATION SUMMARY


,metric,value
0,Total test questions,5
1,Successful live tests,1
2,Quota/error blocked tests,4
3,Live tests passed,1
